1478/check custom script extension usage

ms-sambell opened the pull request:
## PR Summary

Creating a VM and VMSS rule for checking the custom script extension usage for `secureString` functions that aren't the protectedSettings area.

Fixes #1478

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [ ] Unit tests created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

BernieWhite commented on src/PSRule.Rules.Azure/rules/Azure.VM.Rule.ps1:
```diff
@@ -247,3 +247,26 @@ Rule 'Azure.VM.PPGName' -Ref 'AZR-000260' -Type 'Microsoft.Compute/proximityPlac
+    $vmConfig = @($TargetObject);
+
+    if ($PSRule.TargetType -eq 'Microsoft.Compute/virtualMachines') {
+        $vmConfig = @(GetSubResources -ResourceType 'extensions' );
```
Try both `extenstions` and `Microsoft.Compute/virtualMachines/extensions` since both are valid.

BernieWhite commented on tests/PSRule.Rules.Azure.Tests/Resources.VMSS.json:
```diff
@@ -113,7 +113,9 @@
                                 "typeHandlerVersion": "2.0",
-                                "settings": {}
+                                "settings": {
+                                    "commandToExecute": "[concat('Write-Output ', secureString('adminPassword'))]"
```
Is `secureString` a function?

ms-sambell commented:
Halting work on this until this bug is fixed: https://github.com/Azure/PSRule.Rules.Azure/issues/1687

BernieWhite commented on docs/en/rules/Azure.VM.ScriptExtensions.md:
```diff
@@ -0,0 +1,74 @@
+
+### Configure with Azure template
+
+ARM
```
Maybe something like:

```suggestion
To deploy VM extensions that pass this rule:

- Set any secure values within `properties.protectedSettings`.
```

BernieWhite commented on docs/en/rules/Azure.VM.ScriptExtensions.md:
````diff
@@ -0,0 +1,74 @@
+```
+
+### Configure with Bicep
+
````
```suggestion

To deploy VM extensions that pass this rule:

- Set any secure values within `properties.protectedSettings`.

```

BernieWhite commented on docs/en/rules/Azure.VMSS.ScriptExtensions.md:
```diff
@@ -0,0 +1,78 @@
+
+## RECOMMENDATION
+
+Recommendation
```
Probably need to set something here similar to VM.

BernieWhite commented on docs/en/rules/Azure.VM.ScriptExtensions.md:
```diff
@@ -0,0 +1,74 @@
+
+## RECOMMENDATION
+
+Move the script in the `properties.settings` property to the `properties.protectedSettings` property.
```
```suggestion
Consider specifying secure values within `protectedSettings` to avoid exposing secrets during extension deployments.
```

BernieWhite commented on src/PSRule.Rules.Azure/rules/Azure.VM.Rule.ps1:
```diff
@@ -247,3 +247,26 @@ Rule 'Azure.VM.PPGName' -Ref 'AZR-000260' -Type 'Microsoft.Compute/proximityPlac
 #endregion Proximity Placement Groups
+
+# Synopsis: Protect Custom Script Extensions commands
+Rule 'Azure.VM.ScriptExtensions' -Ref 'AZR-000290' -Type 'Microsoft.Compute/virtualMachines', 'Microsoft.Computer/virtualMachines/CustomScriptExtension', 'Microsoft.Compute/virtualMachines/extensions' -Tag @{ release = 'GA'; ruleSet = '2022_09' } {
```
Will need to bump the `ref` and `ruleSet`.

BernieWhite commented on src/PSRule.Rules.Azure/rules/Azure.VMSS.Rule.ps1:
```diff
@@ -50,6 +50,23 @@ Rule 'Azure.VMSS.PublicKey' -Ref 'AZR-000288' -Type 'Microsoft.Compute/virtualMa
 }
 
+# Synopsis: Protect Custom Script Extensions commands
+Rule 'Azure.VMSS.ScriptExtensions' -Ref 'AZR-000291' -Type 'Microsoft.Compute/virtualMachineScaleSets', 'Microsoft.Computer/virtualMachineScaleSets/CustomScriptExtension', 'Microsoft.Compute/virtualMachineScaleSets/extensions' -Tag @{ release = 'GA'; ruleSet = '2022_09' } {
```
Will need to bump the `ref` and `ruleSet`.

BernieWhite reviewed (changes requested):
@ms-sambell Nice. Double check the comments below.

ms-sambell commented:
I've addressed the PR feedback and cleaned up the PR. I'm still not sure how to leverage the SecretReference that exists in the Export-Template file during the rule evaluation. 

Once this is clear I should be able to complete this. I've pinged you on the other issue (Deployment scopes) for clarification.

BernieWhite reviewed (approved):
@ms-sambell Nice work! I have a few comments below, but otherwise all good to merge.

ms-sambell commented on src/PSRule.Rules.Azure/rules/Azure.VM.Rule.ps1:
```diff
@@ -230,6 +230,28 @@ Rule 'Azure.VM.PPGName' -Ref 'AZR-000260' -Type 'Microsoft.Compute/proximityPlac
+    $vmConfig = @($TargetObject);
+
+    if ($PSRule.TargetType -eq 'Microsoft.Compute/virtualMachines') {
+        $vmConfig = @(GetSubResources -ResourceType 'extensions', 'Microsoft.Compute/virtualMachines/extensions' );
```
I missed that. Good call out. 

I believe your code suggestion handles that (thank you).
